### ISSUE TITLE
feature request: Safari app extensions support

### ISSUE THREAD

**Issue by luzik** [keepassxreboot/keepassxc-browser#294]
Please consider providing support for Safari web browser.

**Comment by droidmonkey** [comment]
Not until Safari supports the Web Extensions API. Please see https://stackoverflow.com/questions/47435082/will-safari-ever-support-webextensions-api

**Comment by varjolintu** [comment]
Let's reopen this as Apple just announced that they will support the WebExtension API with the upcoming Safari release. It's unclear if it's only available for the next OS version or for all.

**Comment by varjolintu** [comment]
https://hacks.mozilla.org/2020/06/welcoming-safari-to-the-webextensions-community/ - According to this, Mojave, Catalina and Big Sur are the supported macOS versions.

**Comment by varjolintu** [comment]
Plus it's confirmed that Native Messaging will be supported:
https://developer.apple.com/documentation/safariservices/safari_web_extensions/messaging_between_the_app_and_javascript_in_a_safari_web_extension

**Comment by varjolintu** [comment]
...but I haven't been able to test it yet because it needs a beta version of Big Sur and XCode both.

**Comment by vyashole** [comment]
@varjolintu Both Xcode 12 and Safari 14 are now available on macOS Catalina.

`xcrun safari-web-extension-converter keepassxc-browser` 
converted the extension into an xcode project with the following warning:
```
Warning: The following keys in your manifest.json are not supported by your current version of Safari. If these are critical to your extension, you should review your code to see if you need to make changes to support Safari:
	css
	browser_action
	all_frames
	options_ui
	applications
	contextMenus
	storage
	scripts
	commands
	notifications
	webRequest
	default_locale
	tabs
	webNavigation
	nativeMessaging
	clipboardWrite
	version
	version_name
	name
	web_accessible_resources
	activeTab
	run_at
	webRequestBlocking
	js
	matches
	manifest_version
	icons
	description
```
I can see that some of these might be a problem for normal functioning. Hope you can at least get simple auto fill to work though. All the best.

**Comment by varjolintu** [comment]
@vyashole I haven't tried it out yet, but seems strange if `nativeMessaging` gives a warning because it should be supported.

**Comment by vyashole** [comment]
@varjolintu the warning says it is not supported by the current version of safari i have. Maybe I'm missing something here. 

Native messaging is supported according to https://developer.apple.com/documentation/safariservices/safari_web_extensions/messaging_between_the_app_and_javascript_in_a_safari_web_extension

Anyway it is past midnight where I live. I'll go sleep now. Will check it out tomorrow.

**Comment by cschlosser** [comment]
```
App Name: KeePassXC-Browser
App Bundle Identifier: com.yourCompany.KeePassXC-Browser
Language: Swift
Is this correct? [yes]: 
Warning: The following keys in your manifest.json are not supported by your current version of Safari. If these are critical to your extension, you should review your code to see if you need to make changes to support Safari:
	notifications
	webRequestBlocking
	applications
```
That's what it's outputting for me with Safari (Version 14.0 (15610.1.28.1.9, 15610)) with Xcode Version 12.0 beta 2 (12A6163b) on 10.15.5.

**Comment by varjolintu** [comment]
<img width="675" alt="Screenshot 2020-09-21 at 23 10 12" src="https://user-images.githubusercontent.com/24570482/93816089-b0693900-fc5f-11ea-8dd6-8efcb48f5ac5.png">

Lets see how this goes.

**Comment by vyashole** [comment]
See: https://developer.apple.com/documentation/safariservices/safari_web_extensions/messaging_between_the_app_and_javascript_in_a_safari_web_extension
 > Safari ignores the `application.id` parameter and only sends the message to the containing macOS app’s native app extension. 
This means the safari extension will have to be bundled with either KeePassXC or a standalone proxy.

**Comment by varjolintu** [comment]
> See: https://developer.apple.com/documentation/safariservices/safari_web_extensions/messaging_between_the_app_and_javascript_in_a_safari_web_extension
> 
> > Safari ignores the `application.id` parameter and only sends the message to the containing macOS app’s native app extension.
> 
> This means the safari extension will have to be bundled with either KeePassXC or a standalone proxy.

True. A proxy would be the correct place for it, and I'm gonna do the implementation soon.

**Comment by vyashole** [comment]
Hi @varjolintu 
Have you found time to work on this yet? I'm not asking for an ETA, but just letting you know that I am very interested in this feature, and while I can't be very useful in actual coding, I'll be glad to help out by testing, sharing logs, etc. Let me know in case any help is required.

**Comment by varjolintu** [comment]
@vyashole I have problems compiling the code (event the example one, Native Messaging Demo Extension). Even when I already have latest Catalina, XCode 12 and Safari 14, the code gives me `Cannot find SFExtensionMessageKey in scope` error and I have no idea how to solve it.

**Comment by varjolintu** [comment]
I've been looking at the possible implementation and it seems the best solution for this is to wrap an own application for macOS that contains the extension + the Safari code. In this way we can provide a new Safari Web Extension version for each release. Otherwise we'd need to release a new KeePassXC version for macOS for each extension update, and that is not gonna work.

**Comment by droidmonkey** [comment]
They always gotta be special....

**Comment by varjolintu** [comment]
> They always gotta be special....

Indeed. I almost thought this thing would be easy to do. Basically we have to write an own "proxy" application that works only with Safari.

**Comment by rennefJ** [comment]
> Basically we have to write an own "proxy" application that works only with Safari.
I am not a developer, but I have reseached this topic a little bit. 
How I understand it, the extension needs to be distributed as an app. But the app does not necessarily need to contain any functionality. It can be more ore less an empty shell that just contains the extension. 
But it could also be in included in the main Keepassxx App. Of course if you want to be able to release the extension separately I would also go for the empty shell.

**Comment by rennefJ** [comment]
I found a guy who has created a Safari extension that just adds a reload button.
He has written about it, maybe that is helpful.
[https://hypercritical.co/safari-reload-button/](https://hypercritical.co/safari-reload-button/)

**Comment by droidmonkey** [comment]
@rennefJ this is a _native messaging_ problem, not an extension problem.

**Comment by varjolintu** [comment]
Luckily the XCode project loads the extension source files as references so there's no need to copy files around.

**Comment by OsapBender** [comment]
hey guys. can i home to have support keepassxc in safari soon? (at least in big sur)

**Comment by varjolintu** [comment]
@OsapBender Cannot promise anything yet. I have some troubles sending messages from extension to the application. And I haven't had time to do or test it any further.

**Comment by joconcepts** [comment]
Safari support would be really nice! It will maybe make me move to safari from chrome on macos.

**Comment by Smurk** [comment]
Looking forward for the Safari support as well!

**Comment by joconcepts** [comment]
Any news on this topic?

**Comment by varjolintu** [comment]
@joconcepts Not yet. But we haven't forgotten this one.

**Comment by joconcepts** [comment]
well I need to ask again: any news regarding this integration? :)

**Comment by varjolintu** [comment]
@joconcepts No. We are waiting for feedback from Apple for a certain issue.

**Comment by ngdio** [comment]
Actually, it might be a better idea to implement the macOS Autofill function instead in KeePassXC, which was introduced with Big Sur. Strongbox has implemented this already and it works rather well in Safari even without a browser extension, and possibly even in other applications (haven't tested that).

**Comment by varjolintu** [comment]
@ngdio That's one possibility, yes. Gotta inspect how it works.

**Comment by gagarine** [comment]
Another way would be to create a Swift Xcode project that expose the KeePassXC password to macOS Autofill function.

**Comment by 1oo7** [comment]
> @ngdio That's one possibility, yes. Gotta inspect how it works. The difficulty here is that KeePassXC is not a XCode project. We always have to combine C++ with Objective-C code and that's not a perfect way to do things, and it's a bit slower.

...and you were unable to simply make an Obj.C++ project in Xcode?

**Comment by varjolintu** [comment]
@1oo7 Sadly, it's not that simple.

**Comment by elrazez** [comment]
hello, been reading this interesting topic and wondering, has there been any luck or progress yet? will we have the extension for safari soon? if not, how long you think it will take? approximately.
really appericiate your work. this is my 5yo dream!

**Comment by wkoncki** [comment]
> Actually, it might be a better idea to implement the macOS Autofill function instead in KeePassXC, which was introduced with Big Sur.

I would actually advise again this. Autofill extensions have this rather weird workflow on macOS where you first have to select it from dropdown, then select your password, and then you get info from Safari asking you to save it in its own keychain. It's awkward, and generally inferior to how KeePassXC extension works in Firefox. There is a reason other password managers are not embracing it as well.

**Comment by janprzy** [comment]
I cannot confirm this. I use Strongbox, which has implemented the Autofill API, and it works pretty similar to the iCloud Keychain.
The login credentials are suggested immediately and there are no extra steps involved. I only have to manually select Strongbox if the correct credentials are not offered in the dropdown.

<img width="668" alt="Bildschirmfoto 2021-04-24 um 11 32 49" src="https://user-images.githubusercontent.com/41625296/115954397-13ba9280-a4f1-11eb-964a-6a467a56592e.png">

**Comment by varjolintu** [comment]
> hello, been reading this interesting topic and wondering, has there been any luck or progress yet? will we have the extension for safari soon? if not, how long you think it will take? approximately.
> really appericiate your work. this is my 5yo dream!

Sorry for not bringing any news to this feature sooner. I've been quite busy. Anyway, I had constant problems of debugging the messages sent from the browser extension, and after changing many emails with Apple's support they weren't able to provide any help. What this means for developers is that there's no way to debug any messages sent from Native Messaging API to the Swift side of the code. They suggested that macOS Console application could show the messages, it didn't.

So, because there's no way to actually debug the application side code there's only two choices left: try to continue the implementation blind without any debugging (which takes even more time because I'm not a Swift expert), or we'll make a switch to the Autofill API. The latter needs to be investigated if it's possible to implement in directly to the KeePassXC application which is not a XCode project.

**Comment by sirpiotrek** [comment]
Any news on this issue?

**Comment by varjolintu** [comment]
@sirpiotrek I'll update any news to this thread.

**Comment by joconcepts** [comment]
Any news?

**Comment by SommerEngineering** [comment]
Thanks @varjolintu for not giving up and continuing to try to implement our favorite plugin for Safari. Thanks for the time you invest ❤️ 👍

**Comment by varjolintu** [comment]
@joconcepts There hasn't been any progress lately. I'm tempted to try the new AutoFill API but currently I'm unable to update to Big Sur, which is the requirement.

**Comment by Minims** [comment]
@varjolintu Too old Mac ? if yes have a look at this : https://www.youtube.com/watch?v=znlhI6f7x1Q , it might solve your issue in most of the case.

**Comment by varjolintu** [comment]
> @varjolintu Too old Mac ? if yes have a look at this : https://www.youtube.com/watch?v=znlhI6f7x1Q , it might solve your issue in most of the case.

Not too old, but for multiple reasons I cannot upgrade yet. I can't take any risks with incompatibilities with certain applications and plugins at this point.

**Comment by Tobisaninfo** [comment]
Maybe this plugin is a kind of inspiration for handling the safari difficulties. I use this for some months with safari, but I would like to come back to KeePassXC when a safari integration exists

https://github.com/keeweb/keeweb-connect

**Comment by Immanion** [comment]
Hiya,

Just thought I’d post here to say that about a week ago, 1password removed some functionality from the iOS app around shared sheets; with that and their roadmap to force vaults onto their servers (which I only just learned about), removal of a standalone licence, and force to a paid subscription, you might be seeing an increase in interest as people like me are looking for viable alternatives to 1password. A safari extension is going to be pretty essential to a lot of us looking to switch!

Cheers!

**Comment by grovolis** [comment]
I think switching to the autofill API (like Strongbox) is the only viable solution.I wasn't even aware Apple had exposed such functionality up until recently, it works really well in 99% of cases.

**Comment by svoboda77** [comment]
There is an easy workaround for Safari. Install "userscripts" extension (https://github.com/quoid/userscripts), use this script to inject URL to the tab title:

`document.title = document.title+' - '+document.location;`

This enables autotype hotkey funcionality.

**Comment by n-stone** [comment]
Hey, is some actively looking into this? @varjolintu did you get a chance to upgrade to Big Sur?

**Comment by varjolintu** [comment]
@n-stone I have upgraded to Big Sur yes. I'm gonna start some implementation testing later this month.

**Comment by varjolintu** [comment]
Just a small update: I already have a working example code for the Safari WebExtension. Still needs some work because the socket must be made under a certain group instead of the user's temp folder. But the basic socket stuff already works. It's still unsure what features will actually work after that.

Swift gives me a light headache.

**Comment by n-stone** [comment]
> I cannot confirm this. I use Strongbox, which has implemented the Autofill API, and it works pretty similar to the iCloud Keychain. The login credentials are suggested immediately and there are no extra steps involved. I only have to manually select Strongbox if the correct credentials are not offered in the dropdown.
> 
> <img alt="Bildschirmfoto 2021-04-24 um 11 32 49" width="668" src="https://user-images.githubusercontent.com/41625296/115954397-13ba9280-a4f1-11eb-964a-6a467a56592e.png">

Just my two cents regarding the Autofill API:
I tried it with Strongbox for a couple of weeks now and what really annoys me, is that you can't deactivate apple keychain. Keychain will always be the one to generate passwords and also asks you on new websites if you want to save them in keychain. So there is no way of saving new credentials in strongbox directly. Also the "new credentials popups" can only be deactivated if you turn off the Autofill API itself. 
In the end using the Autofill API is be better than nothing, but compared to a KeepassXC browser extensions it lacks many features.

**Comment by ngdio** [comment]
> I tried it with Strongbox for a couple of weeks now and what really annoys me, is that you can't deactivate apple keychain. Keychain will always be the one to generate passwords and also asks you on new websites if you want to save them in keychain.

Yes, that is definitely an annoyance, but:

> So there is no way of saving new credentials in strongbox directly.

That just a missing feature in Strongbox. It would technically be possible to create an option to create a new entry in the popup that appears when you select Strongbox in the autofill list, it has just not been implemented yet.

> Also the "new credentials popups" can only be deactivated if you turn off the Autofill API itself.

There is no way of doing that for all websites but if you just press "Never save" on every website those popups will not appear again, at least for the websites where you did press it. Again, rather annoying, but a decent workaround.

> but compared to a KeepassXC browser extensions it lacks many features.

I'm actually much more satisfied with the autofill option – it detects much more user and password boxes correctly than the KPXC extension does in other browsers and requires less manual intervention from the user.

**Comment by varjolintu** [comment]
I will also investigate the AutoFill option, but for now it seems the WebExtension way is a little bit easier for me.

**Comment by nicerloop** [comment]
How can we provide help in resolving the issues with Safari WebExtension?

**Comment by varjolintu** [comment]
@nicerloop Not really. I already know what to do, just haven't had enough time to finish these tasks. Thanks for asking.

**Comment by Neavcen** [comment]
@varjolintu Reading this thread is like a blockbuster 😄 still hope for the Happy End! Thanks for spending the time trying to solve this ❤️🙌🏼

**Comment by nicerloop** [comment]
Hi, any news, or any way to provide help?

**Comment by varjolintu** [comment]
@nicerloop The connection works now (so retrieving the correct path + group settings is ready), but then I saw the data is not in correct format when sending it through socket.. so gotta do the deserialization of JSON data again.

So if anyone has a ready idea how to deserialize JSON data in `CVarArg` to a string or `UInt8` array in Swift, hit me with a message :)

**Comment by vit9696** [comment]
CVarArg is just an argument of unknown type in the prototype, like C va_list. Could you provide some context please?

**Comment by varjolintu** [comment]
@vit9696 Just updated the branch. It's way behind the current develop but it works with a modified KeePassXC version (not available yet). The main code is here:
https://github.com/keepassxreboot/keepassxc-browser/blob/feature/safari_support/keepassxc-browser-safari/KeePassXC-Browser/KeePassXC-Browser%20Extension/SafariWebExtensionHandler.swift

I had a lot of trouble to get the JSON stuff work properly. So for now I just parse the raw string so at least I can get some response from KeePassXC side. Gotta make a branch for KeePassXC where the Safari support is enabled so others can test this too.

**Comment by vit9696** [comment]
@varjolintu, I gave a brief look to the codebase you mentioned.

To be honest, I both failed to understand what happens in the code you provided and what needs to happen, because there is no specification for `beginRequest()` function, like not even a single comment :D Could you please explain:

- what do you give to that function?
- what do you actually get in this function?
- what does the function currently do?
- what does the function need to do?

In addition to not having these explained there is currently no easy way to test the setup due to apparent KeePassXC modifications. When I downloaded the codebase, it also did not build for me. I include the patches I needed to apply here: [patches.zip](https://github.com/keepassxreboot/keepassxc-browser/files/8443996/patches.zip).

Even so, I made the following assumption:

- JavaScript code calls `browser.runtime.sendNativeMessage` with the following payload:

     ```json
      {"message": {"action": "….", "publicKey": "….", "nonce": "….", "message": "…." }}
     ```

     The values transferred conform to the [protocol](https://github.com/keepassxreboot/keepassxc-browser/blob/develop/keepassxc-protocol.md), and confusingly there are _two_ `message` fields here. One for Safari API, another for the protocol.

- `beginRequest` gets a list of `inputItems` of size 1 (contains `message`), which `userInfo` is a dictionary containing key value pairs `action`, `message`, etc.

- `beginRequest` needs to forward `userInfo` to a socket of an upstream target (I assume KeePassXC) in JSON format.

If this assumption is correct, I can imagine two routes to implement this.

1. Transfer a string rather than a dictionary via the Safari API on the JS side and forward it to whatever proxy you normally have for other browsers. 
2. Transfer the dictionary itself (probably cleaner, if you end up writing the KeePassXC interaction code directly in Swift) and reencode whatever is necessary on the Swift side.

Since, I do no regularly write in Swift, and without an ability to test this code I cannot guarantee that the suggested code snippets will work out of the box, but this is how I believe it can be implemented for both routes:

1. String transfer
    1. Replace:
     ```js
     browser.runtime.sendNativeMessage(keepass.nativeHostName, { message: request) });
     ```
     with

     ```js
     browser.runtime.sendNativeMessage(keepass.nativeHostName, { message: JSON.stringify(request) });
     ```

     2. Use something like:

     ```swift
	func beginRequest(with context: NSExtensionContext) {
        guard let item = context.inputItems.first as? NSExtensionItem else {
            context.cancelRequest(withError: NSError())
            os_log(.error, "Invalid amount of arguments")
            return
        }
        
        guard let str = item.userInfo?[SFExtensionMessageKey] as? String else {
            context.cancelRequest(withError: NSError())
            os_log(.error, "Invalid message receieved")
            return
        }
     ```

2. Dict transfer

```swift
guard let item = context.inputItems.first as? NSExtensionItem else {
    context.cancelRequest(withError: NSError())
    os_log(.error, "Invalid amount of arguments")
    return
}

let jsonData = try? JSONSerialization.data(withJSONObject: item.userInfo?[SFExtensionMessageKey] as Any)
if (jsonData == nil) {
    os_log(.error, "JSON serialization error")
    return
}

let string = String(data: jsonData!, encoding: String.Encoding.utf8)
if (string == nil)
{
    os_log(.error, "Error parsing JSON data")
    return
}
```

**Comment by varjolintu** [comment]
@vit9696 Ah, the code is very crude and lacks comments :D Sorry for that. The function is this one:
https://developer.apple.com/documentation/foundation/nsextensionrequesthandling/1413395-beginrequest

Basically it needs to do what `keepassxc-proxy` does:
1. Get the JSON message from the extension
2. Pass it to the Unix Domain socket KeePassXC is listening
3. Wait for response from the socket
4. Return the message as JSON back to the extension

**Comment by vit9696** [comment]
@varjolintu https://github.com/vit9696/keepassxc-browser/commit/37331a1c9e56b34ffd78c2ab49cd9ae43f593fc4 this commit fixes step 1 for me and will likely work for step 2 for you. As you can see the string containing json-encoded message is properly received and can be printed to `Console.app`.

I did not have your modified KeePassXC implementation, so I stopped here, yet I suppose as long as the socket is connected the main application should receive the message properly.

**Comment by vit9696** [comment]
I additionally fixed the memory handling issue in the sending code and avoided unnecessary copying in https://github.com/vit9696/keepassxc-browser/commit/1d82d468cd007c02a83997ab8f1999aaee2fc9da. At the very least I see messages being received, waited, and replied back via netcat unix sockets. I believe this should be enough for it to work, although there is no way for me to test it without the .app.

**Comment by varjolintu** [comment]
@vit9696 Thank you! I'll check out the changes today.

**Comment by vit9696** [comment]
@varjolintu, I may be misunderstanding something, but I have a suspect that your implementation of `keepass.sendNativeMessage` in JavaScript might be incompatible with Safari. Whenever I set a breakpoint to line 630, it is always hit when I press the "reload" button, and I see the message transferred after I step over 631, however, the promise is never awaited:

<img width="1778" alt="Screenshot 2022-04-17 at 08 13 59" src="https://user-images.githubusercontent.com/4348897/163701441-958c4c38-6abc-46fc-a1e8-4b3e86515d99.png">

But printing the promise itself works fine:


<img width="1396" alt="Screenshot 2022-04-17 at 08 24 43" src="https://user-images.githubusercontent.com/4348897/163701811-0c090b99-ea10-4625-866a-d8c63f94a4f6.png">


REF: https://developer.apple.com/documentation/safariservices/safari_app_extensions/passing_messages_between_safari_app_extensions_and_injected_scripts
REF: https://developer.apple.com/documentation/safariservices/safari_web_extensions/troubleshooting_your_safari_web_extension

**Comment by varjolintu** [comment]
@vit9696 I noticed the same. Just updated to sources rebased to the current develop branch. Apple documentation also uses a third parameter for the response with `browser.runtime.sendNativeMessage()` but that actually gives an error in the background scripts. It's possible that the polyfill library interferes with this, but this needs to be verified.

**Comment by varjolintu** [comment]
@vit9696 The polyfill library was the problem. Using `chrome.runtime.sendNativeMessage()` instead bypasses the library and I was able to get the key change to be successful. Committing the fix soon.

**Comment by varjolintu** [comment]
And now that the basic synchronized connection works, the code needs to be altered in a way we listen to socket messages all the time (in a separate thread if possible) and pass the responses to the extension side. Now all broadcast messages (like database unlocked) are not identified at all.

**Comment by vit9696** [comment]
Good, congrats. How does async connection work on JS and KeePassXC sides? Can have a look, but I need to get an idea first.

**Comment by varjolintu** [comment]
@vit9696 KeePassXC basically listens to the local socket's signal `readyRead()` -> every time something is available for reading, KeePassXC handles the response and sends it back.

Swift side in short:
- Connect to the socket and listen any messages
- When JS side sends a JSON message, pass it to the socket via `beginRequest()`-
- When the listener gets a response, pass it from the Swift side to the JS side.

KeePassXC can also send "broadcast" messages to all connected clients/proxies, and it assumes those are listening. I haven't investigated yet how the Swift side listener could work. If possible, I'll try to test everything I can with the current implementation.

**Comment by vit9696** [comment]
So, basically, the current part that is missing is:

- Listen on the socket for any message from KeePassXC
- If it landed, send a message to the JavaScript code

Does JavaScript allow that? From the docs it looks like it imposes a request/reply schema…

**Comment by varjolintu** [comment]
@vit9696 This is used for listening to broadcast messages:
https://github.com/keepassxreboot/keepassxc-browser/blob/develop/keepassxc-browser/background/client.js#L340=

**Comment by vit9696** [comment]
@varjolintu so far I have not seen an ability to send random message from Safari Extension in Swift to Safari Extension in JavaScript. You might have to poll the Swift side for such messages.

**Comment by vit9696** [comment]
https://github.com/vit9696/keepassxc-browser/commit/21b2349f02efe9c71b3f66ecbc737c2154b5b0c9 this seems to work for me, but you will have to adapt the message receiving code to handle the fields correctly.

**Comment by shuxiao9058** [comment]
good news, how is going with this feature now?

**Comment by varjolintu** [comment]
Just hang on. Like said before, I'll post updated when I have something new.

**Comment by DShirk** [comment]
@varjolintu thanks for all your efforts
Contribution made on open collective

**Comment by mietzen** [comment]
> Just hang on. Like said before, I'll post updated when I have something new.

I know it's done when it's done, but did you discover any problems? It seemed like you where making great progress in April or did you just have no time for this project? Again thanks for effort!

**Comment by varjolintu** [comment]
@mietzen Not enough time for everything :) 1.8.x releases took a big effort, along with Manifest V3 changes, Webauthn investigation etc.. I'll get back to this when I know I have a longer time period to focus just on one thing. Thanks!

**Comment by varjolintu** [comment]
"Outgoing Network Connections" entitlement needed to be added to the sandbox so Unix Domain sockets can be connected. My old computer didn't have such requirement, so I was stuck for a moment. Also changed the App Group to `org.keepassxc.KeePassXC` instead of just `KeePassXC`.

**Comment by varjolintu** [comment]
@vit9696 Btw, I tried your socket polling code and it works, but only when you open the popup or somehow else activate the message transfer with the extension. Currently I'm having hard time to figure out where and how to make a passive version of it that always runs in the background. Any ideas?

**Comment by gencys** [comment]
> @vit9696 Btw, I tried your socket polling code and it works, but only when you open the popup or somehow else activate the message transfer with the extension. Currently I'm having hard time to figure out where and how to make a passive version of it that always runs in the background. Any ideas?

Hey there, I am just butting in real quick and maybe you know all of this already but here goes.

With browser extensions there are 3 scripts that can run:
* Content scripts, i.e. a script that is injected on the webpage and will only run on pages specified in the manifest.
* Popup scripts, i.e. the scripts that you can use in the popup's page of the extension.
* Background scripts, as the name suggests those scripts run in the background.

And all of those scripts can communicate together by using the message API. From the last few messages of the discussion my guess is that you're searching for those background scripts. However there has been a change since manifest v3 which make background scripts less powerful: they can't be persistent anymore... But there is something else that could be useful: service workers. Those work as persistent background scripts. Here are some links you may find useful: [on background scripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts), [on service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers) and [some apple doc](https://developer.apple.com/documentation/safariservices/safari_web_extensions/optimizing_your_web_extension_for_safari).

I don't know what you have on your mind for this extension. But I could see things working with only a non-persistent background script. I hope that was at least a little bit useful :)

Good luck !

**Comment by varjolintu** [comment]
@gencys Thanks. Gotta check if using V3 with a service worker could solve my problem. Luckily there's already a PR for Manifest V3 so I can just copy that one to the Safari extension.

**Comment by Daniel-Kitz** [comment]
So, how is it going? Can we maybe help with anything?

**Comment by ntinti** [comment]
Any news? Can I help testing? I want to kick my 1Password... :-)

**Comment by davidkessler-ch** [comment]
@varjolintu Is this still being worked on? I may get time to look into it, I have worked on Safari extensions before.. Is #1035 up to date?

**Comment by varjolintu** [comment]
@davidkessler-ch I have hit a dead end, and don't know how to continue.

**Comment by nicerloop** [comment]
Hi, what dead end did you hit?

**Comment by varjolintu** [comment]
> Hi, what dead end did you hit?

The background script that listens to socket messages is only active when doing something. It's not working as a passive script.

**Comment by nicerloop** [comment]
If I understand the protocol (https://github.com/keepassxreboot/keepassxc-browser/blob/develop/keepassxc-protocol.md) correctly, it is only relevant for database-locked and database-unlocked signals. Are these absolutely required, or could'n they be replaced with some sort of status polling or on-demand calls in case of locked database ?

**Comment by varjolintu** [comment]
> If I understand the protocol (https://github.com/keepassxreboot/keepassxc-browser/blob/develop/keepassxc-protocol.md) correctly, it is only relevant for database-locked and database-unlocked signals. Are these absolutely required, or could'n they be replaced with some sort of status polling or on-demand calls in case of locked database ?

Those signals are also sent when switching to another database. Maybe when Protocol V2 is implemented, these kind of things are no longer relevant. It might be that a polling function is needed. Gotta experiment some.

**Comment by OperativeThunny** [comment]
is there something I can do to help with this issue? I have a MacBook and Xcode and know how to code

**Comment by clemlesne** [comment]
I strongly vote for this feature :)

**Comment by mothepro** [comment]
Using @varjolintu's great work as a launchpad; I will attempt to add support for safari on MacOS in #2295.

**Comment by penguin359** [comment]
I am interested in helping. I have recently been experimenting with porting other, less complex WebExtensions to Safari and would like to see this one ported next. What is the current status of it and what still needs to be worked on? Do you have a GitHub Project or Issue tracking the list of to-do items?

**Comment by varjolintu** [comment]
@penguin359 https://github.com/keepassxreboot/keepassxc-browser/pull/1035 The status is currently stalled, because I haven't had time to look at this any further.

**Comment by coignard** [comment]
Any news?

**Comment by benjuade** [comment]
Bitwarden and Netnewswire somehow were able to create an extension automatically while installing the app:

<img width="836" height="592" alt="Image" src="https://github.com/user-attachments/assets/de596d0d-a4af-4083-b8ff-032a9f8786f0" />

**Comment by varjolintu** [comment]
@benjuade Yes, because Bitwarden has an extension for Safari..

**Comment by varjolintu** [comment]
> Any news?

I created a new PR for this. For the problematic native messaging, I decided to rely on the experimental WebSocket support that I implemented earlier for testing an alternative connection method for Flatpak browsers. It seems it works very nice with Safari as well :) I'll continue the implementation from that.

**Comment by aaszodi** [comment]
I am just a user, I cannot help with the coding.

However... I have found a macOS/iOS password manager called ["Keepassium".](https://keepassium.com). I tried out the free version, it works flawlessly with Safari. And it is [open source](https://github.com/keepassium/KeePassium) . Now: would it be possible for the KeepassXC developers to look into that codebase and use it as inspiration? Or is it against the license/is unethical/whatever...?

Please forgive my lack of experience in these matters. Thank you.

**Comment by varjolintu** [comment]
@aaszodi It's a totally different thing.

**Comment by x144k** [comment]
don't know if this may be useful for someone, but thought I would share it here.
hoping keepassxc gets this sorted out natively, sooner than later.

https://www.griesslehner.at/safarikeepassxc.html

**Comment by varjolintu** [comment]
@x144k Please don't share or promote closed-source alternatives here.

**Comment by x144k** [comment]
> [@x144k](https://github.com/x144k) Please don't share or promote closed-source alternatives here.

Understood, sorry about that. I didn't realize it was closed-source. I was just trying to help people who've been waiting on Safari support, but I get why you can't have unaudited closed-source tools pointed to from here. Looking forward to official support whenever it lands.

**Comment by x144k** [comment]
> [@x144k](https://github.com/x144k) Please don't share or promote closed-source alternatives here.

Good news! With a little reasoning, I managed to influence the dev to release the source code! He's made the leap to open-source. This should be great news for many users who've been waiting for so long for their favorite password manager to be supported in Safari. Perhaps you guys can use the logic as inspiration for your efforts here? Regards.

https://git.griesslehner.at/markus/SafariKeePassXC

**Comment by varjolintu** [comment]
@x144k That page is not found.

**Comment by x144k** [comment]
> [@x144k](https://github.com/x144k) That page is not found.

it was working moments ago. But now he created a GitHub repo:

https://github.com/mgriesslehner/SafariKeePassXC
